### ISSUE TITLE
How to test an endpoint that depends of an asynchronous processing

### ISSUE THREAD

**Issue by wedla** [pact-foundation/pact-jvm#1561]
Hello, guys! I'm opening this issue after browsing the docs and the opened and closed issues and not finding what I wanted for. Please excuse me if I didn't look up correctly.

I'll explain the flow: I have an API with two endpoints. In the first one, I send a body that represents an order service. This order is then processed asynchronously by a consumer application that verifies if it is eligible for a benefit. The second endpoint is a GET that returns to me the status of the elegibility of that order service. 

The consumer application that proccesses asynchronously the order services gets them from a kafka topic.

I have two different test classes: the one that tests the sending of the order services and the second that tests the elegibility of them. By the business rules, I know exactly the status of all of them, but the tests execute fastly, so sometimes I get a 404 not found because the order is not totally processed yet. I'm using Thread.sleep() to wait for this processing, but it's really bad, because sometimes I have to wait one minute, and other times I have to wait for two, three minutes.

I want to know how can I test this second endpoint without the use of explicit waits. I wanted something like implicit wait or a timeout I can set, so the test will fail after the configured timeout.

Is this feasible in Pact, my friends?! I appreciate since now all the help you can give me.

**Comment by wedla** [comment]
Some info that can help you guys understand my "pain".

I'm not using a Pact Broker. I have a father project that contains a consumer and a provider children. The pact files are generated in the consumer and then are copied to a dir in the provider by using the antrun plugin.

The first class that is executed in the provider is the one who transmits the order services, and then imediatally after it the other that verifies the elegibily status. There's no time to the asynchronous application to proccess the orders I sent in the first test class, so the tests in the second class fail.

I'm not testing the producer-consumer Kafka behaviour. These are two HTTP endpoints that have contracts, so this type of test applies to them.

I was thinking of using JUnit @BeforeClass annotation in the second test class to wait for X minutes (using Thread.sleep()), so the consumer app has enough time to proccess the orders, but I don't think is a good solution.

Still wanting your help, guys. Since now, I appreciate all your attention, and want to say that I love to work using this framework.

**Comment by rholshausen** [comment]
The problem is you are trying to test two interactions together in sequence like you would with a functional or integration test. Pact is designed to test that the interactions are correct in isolation.

The two tests should not be coupled together. The first test should test: have I sent the correct request to the order service, and did I get the correct response back from it.

The second test should be: Given that the order has been processed and the message is available on the Kafka topic, do I get the correct status response.

You can use a combination of provider states and mocking to setup the correct behavior for the tests to work like that.

**Comment by wedla** [comment]
Hi, Ronald, thank you so much for your answer. Gentle as always.

Actually, we have found that our development solution is not good. If ten order services are sent, the consumer application will proccess them in a FIFO way, and this could take minutes. We don't have a multithreading proccessing, but a mono one. The solution is back to our architecture team so they can elaborate better.

But I still want to know how could I handle this in Pact, because I have two isolated classes, as you said. The first one sends the service orders, and the second one gets its elegibility status. Supposing that I still have some time to proccessing, instead of minutes, seconds, how could I put a wait between the execution of the two classes? Example: I want to execute the second class 30 seconds after the first one? Is this possible? How can I do this in the most elegant way?

**Comment by rholshausen** [comment]
I don't think you should execute them as separate tests, as they are dependent one each other. There may be JUnit extensions that help with this, but I would execute them as a single test. If you don't want to use waits for the second interaction, you may need to use some type of polling mechanism.
